core/src/services/onedrive/pager.rs

```diff
@@ -0,0 +1,160 @@
+    },
+    EntryMode, Metadata,
+};
+
+use super::{
+    error::parse_error,
+    graph_model::{GraphApiOnedriveListResponse, ItemType},
+};
+use crate::Result;
+use async_trait::async_trait;
+use http::{header, Request, Response};
+
+pub struct OnedrivePager {
+    root: String,
+    path: String,
+    access_token: String,
+    client: HttpClient,
+    next_link: Option<String>,
+    done: bool,
+}
+
+impl OnedrivePager {
+    const DRIVE_ROOT_PREFIX: &'static str = "/drive/root:";
+
+    pub(crate) fn new(
+        root: String,
+        path: String,
+        access_token: String,
+        client: HttpClient,
+    ) -> Self {
+        Self {
+            root,
+            path,
+            access_token,
+            client,
+            next_link: None,
+            done: false,
+        }
+    }
+}
+
+#[async_trait]
+
+impl oio::Page for OnedrivePager {
+    async fn next(&mut self) -> Result<Option<Vec<oio::Entry>>> {
+        if self.done {
+            return Ok(None);
+        }
+        let response = self.onedrive_get().await?;
+
+        let status_code = response.status();
+        if !status_code.is_success() {
+            if status_code == http::StatusCode::NOT_FOUND {
+                return Ok(None);
+            }
+            let error = parse_error(response).await?;
+            return Err(error);
+        }
+
+        let bytes = response.into_body().bytes().await?;
+        let decoded_response = serde_json::from_slice::<GraphApiOnedriveListResponse>(&bytes)
+            .map_err(new_json_deserialize_error)?;
+
+        if let Some(next_link) = decoded_response.next_link {
+            self.next_link = Some(next_link);
+        } else {
+            self.done = true;
+        }
+
+        let entries: Vec<oio::Entry> = decoded_response
+            .value
+            .into_iter()
+            .map(|drive_item| {
+                let name = drive_item.name;
+                let parent_path = drive_item.parent_reference.path;
+                let parent_path = parent_path
+                    .strip_prefix(Self::DRIVE_ROOT_PREFIX)
+                    .unwrap_or("");
+
+                let path = format!("{}/{}", parent_path, name);
+
+                let normalized_path = build_rel_path(&self.root, &path);
+
+                let entry: oio::Entry = match drive_item.item_type {
+                    ItemType::Folder { .. } => {
+                        let normalized_path = format!("{}/", normalized_path);
+                        oio::Entry::new(&normalized_path, Metadata::new(EntryMode::DIR))
+                    }
+                    ItemType::File { .. } => {
+                        oio::Entry::new(&normalized_path, Metadata::new(EntryMode::FILE))
+                    }
+                };
+                entry
+            })
+            .collect();
+
+        Ok(Some(entries))
+    }
+}
+
+impl OnedrivePager {
+    async fn onedrive_get(&mut self) -> Result<Response<IncomingAsyncBody>> {
```

Comment:
This looks a bit repetitive here.

Maybe we should refactor all `onedrive_get()` into one generic method in the backend struct?

Reply:
I prefer merge `build_request_url` into this API directly.